compiler/rustc_hir_analysis/src/astconv/errors.rs

```diff
@@ -585,6 +586,24 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
             }
         }
 
+        let bound_names = trait_bounds
+            .iter()
+            .flat_map(|poly_trait_ref| {
+                poly_trait_ref.trait_ref.path.segments.iter().flat_map(|x| {
+                    x.args.iter().flat_map(|args| {
```

Comment:
Why are you looking at all of the segments in the trait ref? That shouldn't be necessary. Only the final path segment should be interesting here.

Reply:
Also, please don't `iter()` over an `Option` -- `if let` or `let .. else` should work here more effectively.